### PR TITLE
fix(core): use federation decision_hash for executor idempotency

### LINKED ISSUE #2095: fix(core): use propagated decision_hash as idempotency key for federation decisions
## Context

Follow-up to #2094 (which completed canonical-path `decision_hash` propagation through `FederationEffect::TerminateClearing` / `RevokeVouch`) and #2093 (which bound `decision_hash` into the terminate/revoke provenance hashes). Surfaced by codex review on #2094: https://github.com/InterCooperative-Network/icn/pull/2094#discussion_r3439713890

## Problem

On the production `create_decision_executor_callback` path, the decision idempotency key is derived by `extract_decision_hash` (`crates/icn-core/src/supervisor/decision_executor.rs:822-837`), which **only matches Treasury effects**:

```rust
if let KernelEffect::Treasury(
    TreasuryEffect::Spend { decision_hash, .. }
    | TreasuryEffect::CreateBudget { decision_hash, .. }
    | TreasuryEffect::ReleaseEscrow { decision_hash, .. },
) = effect { ... }
```

For a **federation-only** decision it returns `None`, so the callback falls back to `decision_receipt_id` as the idempotency key (`:722` `extract_decision_hash(&effects).unwrap_or_else(|| decision_receipt_id.clone())`), and `DecisionExecutor::execute` dedupes on that value (`:395` `self.store.get(decision_hash)`).

**Consequence:** in the same-`decision_receipt_id` / different-`decision_hash` scenario (exactly what #2093's provenance binding was meant to distinguish), the second federation decision is skipped at the idempotency layer **before** the federation service binds the newly-propagated hash. So #2094's end-to-end benefit is not realized for that scenario.

## Proposed change

Teach `extract_decision_hash` to also read the `decision_hash` carried by federation effects that now have it — `FederationEffect::SettleClearing`, `TerminateClearing`, `RevokeVouch` — so the idempotency key reflects the actual decision.

## Why this is its own issue (not folded into #2094)

- It **changes the idempotency key** for federation decisions — replay-safety-sensitive, directly adjacent to #2092 (control effect replay safety). It must not ride on a provenance-only PR.
- It is **pre-existing** (the extractor handled only Treasury before #2094 too).
- It needs dedicated replay-safety tests: same-receipt/different-hash must NOT dedupe; same-hash MUST dedupe; empty-hash legacy fallback behavior must stay explicit and safe.

## Acceptance criteria

- [ ] `extract_decision_hash` returns the federation effect's `decision_hash` when non-empty.
- [ ] Replay-safety test: two federation decisions with the same `decision_receipt_id` but different `decision_hash` are both executed (not deduped).
- [ ] Replay-safety test: identical `decision_hash` still dedupes (idempotency preserved).
- [ ] Empty/legacy `decision_hash` falls back to prior behavior (no idempotency regression).
- [ ] No change to non-federation idempotency behavior.

Proposed branch: `fix/federation-idempotency-decision-hash-key`

### PR DESCRIPTION
**Pull request by fahertym**
## Summary

Fixes #2095. Extends decision-executor idempotency-key extraction so federation effects with a non-empty `decision_hash` use that hash as the primary idempotency key instead of falling back to `decision_receipt_id`.

## Context

- #2093 bound `decision_hash` into federation terminate/revoke provenance hashes.
- #2094 propagated `decision_hash` through the canonical `KernelEffect::Federation` path into `FederationOperation` and service requests.
- This PR completes the executor-layer follow-up and preserves replay compatibility across the #2094 → #2096 upgrade window.

Before this PR, federation-only decisions returned `None` from `extract_decision_hash`, so the production callback fell back to `decision_receipt_id`. In a same-receipt / different-decision-hash case, a second federation decision could dedupe before the federation service saw the propagated hash.

## What changed

- `extract_decision_hash` now extracts non-empty `decision_hash` from:
  - `FederationEffect::SettleClearing`
  - `FederationEffect::TerminateClearing`
  - `FederationEffect::RevokeVouch`
- Empty/legacy federation hashes still return `None`, preserving the receipt-id fallback.
- Treasury extraction behavior is unchanged.
- Added replay compatibility for #2094-era records:
  - existing receipt-keyed terminal records remain deduped after upgrade;
  - existing receipt-keyed in-flight/retryable records continue under the same legacy key after upgrade;
  - a legacy receipt-keyed row is only sticky when it represents the same decision.
- Distinct same-receipt federation decisions still keep their own `decision_hash` key.

## Replay / idempotency behavior

- Same receipt id + different non-empty federation hashes no longer dedupe incorrectly.
- Identical federation hash still dedupes.
- Empty/legacy federation hash preserves previous receipt-id fallback.
- Existing receipt-keyed terminal federation records remain deduped after upgrade.
- Existing receipt-keyed in-flight/retryable federation records continue under the same legacy key after upgrade.
- A legacy receipt-keyed record does not collapse a distinct same-receipt/different-hash federation decision.

## What did NOT change

- No authorization change.
- No token issuance change.
- No OpenAPI/SDK change.
- No federation service provenance-formula change.
- No `decision_receipt_id` field added to federation effects.
- No NYCN change.
- No live-federation / production / pilot-readiness claim.

## Tests

TDD was used: tests were written red first, then implementation was narrowed until green.

Unit coverage:
- federation extraction for `SettleClearing`, `TerminateClearing`, and `RevokeVouch`;
- empty-hash fallback;
- distinctness / determinism;
- Treasury precedence unchanged.

Integration coverage through `create_decision_executor_callback`:
- same receipt + different federation hashes do not dedupe incorrectly;
- empty federation hash falls back to receipt id;
- legacy receipt-keyed terminal record dedupes replay after upgrade;
- legacy receipt-keyed in-flight/retryable record remains sticky after upgrade;
- legacy receipt-keyed record does not collapse a distinct same-receipt federation decision.

Local validation:
- `cargo fmt --all --check`
- `cargo test -p icn-core --lib decision_executor` → 18 passed
- `cargo test -p icn-core --test decision_executor_runtime_test` → 23 passed
- `cargo test -p icn-core --lib` → 406 passed
- `cargo clippy -p icn-core --all-targets --all-features -- -D warnings`
- `ops/scripts/drift-check.sh` → PASS

### CODE REVIEW

File: icn/crates/icn-core/src/supervisor/decision_executor.rs [hunk 1]
```diff
@@ -818,19 +818,31 @@ pub fn create_decision_executor_callback_with_sink(
...
 }
 
-/// Extract `decision_hash` from the first effect that carries one.
+/// Extract the governance `decision_hash` from the first effect that carries a
+/// non-empty one, scanning in order. Used as the executor idempotency key.
+///
+/// Both Treasury and Federation effects embed a `decision_hash` (the latter
+/// since #2093/#2094). Empty/legacy hashes are skipped so the caller falls back
+/// to `decision_receipt_id`, preserving backward compatibility for pre-#2094
+/// serialized effects.
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs]: Tightened the doc comment in 83acaeef so it reads "from the first *supported* effect variant" carrying a non-empty `decision_hash`, and named the allow-list explicitly (`TreasuryEffect::{Spend, CreateBudget, ReleaseEscrow}`, `FederationEffect::{SettleClearing, TerminateClearing, RevokeVouch}`). It now also states that other variants and effect families that embed `decision_hash` (the remaining Treasury variants, Membership) are intentionally not consulted here. Wording only — extraction semantics are unchanged in this PR; widening the allow-list would be a separate, deliberate change.

[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs]: Doc comment tightened in 83acaeef to name the supported allow-list explicitly (`TreasuryEffect::{Spend, CreateBudget, ReleaseEscrow}`, `FederationEffect::{SettleClearing, TerminateClearing, RevokeVouch}`) and note other variants/families are intentionally not consulted. Extraction semantics are unchanged. _🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: icn/crates/icn-core/src/supervisor/decision_executor.rs [hunk 2]
```diff
@@ -818,19 +818,31 @@ pub fn create_decision_executor_callback_with_sink(
...
+            KernelEffect::Treasury(
+                TreasuryEffect::Spend { decision_hash, .. }
+                | TreasuryEffect::CreateBudget { decision_hash, .. }
+                | TreasuryEffect::ReleaseEscrow { decision_hash, .. },
+            ) => decision_hash,
+            KernelEffect::Federation(
+                FederationEffect::SettleClearing { decision_hash, .. }
+                | FederationEffect::TerminateClearing { decision_hash, .. }
+                | FederationEffect::RevokeVouch { decision_hash, .. },
+            ) => decision_hash,
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs:916]: Addressed in b2b53446. `DecisionExecutor::execute` now does a read-only backward-compatibility probe: when `decision_hash != decision_receipt_id`, after missing the primary `store.get(decision_hash)` it also checks for a **terminal** record under `decision_receipt_id` and dedupes against it (returning `Ok(vec![])` like the existing terminal-dedupe path, so sink/evidence behavior is unchanged).

- Adds the legacy receipt-key fallback **before** executing under the new federation hash key.
- Preserves the #2095 forward behavior: distinct non-empty federation hashes still execute independently, and the empty-hash case still falls back to the receipt id (the check is a no-op when `decision_hash == decision_receipt_id`).
- Prevents duplicate execution for pre-#2096, #2094-era federation records that were keyed under `decision_receipt_id`.
- No record migration — read-only compatibility shim. Treasury behavior unchanged (treasury decisions were always hash-keyed, so no receipt-keyed record exists for them).

Added an integration test through the public `create_decision_executor_callback` path: it seeds a legacy receipt-keyed terminal `ExecutionRecord`, replays a federation effect carrying a non-empty `decision_hash`, and asserts no second hash-keyed record is created and the legacy record is not re-executed.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: icn/crates/icn-core/src/supervisor/decision_executor.rs [hunk 3]
```diff
@@ -429,6 +429,33 @@ impl DecisionExecutor {
...
+        // above), not find the legacy receipt-keyed record, and re-run the
+        // operation. When the key differs from the receipt id, also probe for a
+        // terminal record under the receipt id and dedupe against it. This is a
+        // read-only compatibility shim — no record migration — and is a no-op for
+        // decisions whose key never moved (empty-hash fallback sets
+        // `decision_hash == decision_receipt_id`; treasury decisions were always
+        // hash-keyed, so no receipt-keyed record exists for them).
+        if decision_hash != decision_receipt_id {
+            if let Some(legacy) = self.store.get(decision_receipt_id)? {
+                if legacy.is_terminal() {
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs]: Addressed in da7130e9. Replaced the terminal-only check with a sticky-legacy-key selection at the top of `DecisionExecutor::execute`: when a record already exists under `decision_receipt_id` and differs from `decision_hash`, the whole call operates under that legacy key.

So an in-flight / retryable legacy record (`Pending` / `Executing` / retryable `Failed`) is now honored — it continues under its own key with retry count preserved, rather than being orphaned while a fresh `decision_hash` record is created with retries reset. Terminal legacy records still dedupe via the same path. Read-only key selection — no record migration. No-op when the keys are equal (empty-hash fallback) or for treasury (always hash-keyed, so no receipt-keyed record exists).

Added a non-terminal replay test (`test_federation_replay_honors_legacy_in_flight_receipt_record`); the terminal-legacy and forward different-hash tests are unchanged and still pass.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: icn/crates/icn-core/src/supervisor/decision_executor.rs [hunk 4]
```diff
@@ -391,6 +391,29 @@ impl DecisionExecutor {
...
+        // already exists under the receipt id (and the keys differ), keep
+        // operating under that legacy key so its full idempotency state (terminal
+        // dedupe, retry count, in-flight status) is honored. This is read-only
+        // key selection — no record migration — and a no-op when the key never
+        // moved: the empty-hash fallback sets `decision_hash ==
+        // decision_receipt_id`, and treasury decisions were always hash-keyed so
+        // no receipt-keyed record exists for them.
+        let decision_hash: &str = if decision_hash != decision_receipt_id
+            && self.store.get(decision_receipt_id)?.is_some()
+        {
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs]: Addressed in de53e1a0. The legacy receipt key is now adopted only when the stored record is the *same* decision — `extract_decision_hash(&legacy.effects).as_deref() == Some(decision_hash)`.

A distinct federation decision that merely shares the `decision_receipt_id` keeps its own `decision_hash` key, so the #2095 same-receipt/different-hash distinctness guarantee holds across the upgrade. The same decision (terminal or in-flight) still dedupes / continues under its legacy receipt key. No-op for treasury (always hash-keyed) and the empty-hash fallback.

Added a regression test (`test_federation_legacy_record_does_not_collapse_distinct_same_receipt_decision`) proving a distinct same-receipt decision executes under its own hash key, and updated the two legacy-replay tests to seed realistic effects carrying the decision hash.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: icn/crates/icn-core/src/supervisor/decision_executor.rs [hunk 5]
```diff
@@ -391,6 +391,38 @@ impl DecisionExecutor {
...
+        // ties adoption to decision identity. Read-only key selection — no record
+        // migration — and a no-op when the key never moved (empty-hash fallback
+        // sets `decision_hash == decision_receipt_id`; treasury was always
+        // hash-keyed, so no receipt-keyed record exists for it).
+        let decision_hash: &str = if decision_hash != decision_receipt_id {
+            match self.store.get(decision_receipt_id)? {
+                Some(legacy)
+                    if extract_decision_hash(&legacy.effects).as_deref() == Some(decision_hash) =>
+                {
+                    decision_receipt_id
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs:429]: Addressed in c64b9f69. `DecisionExecutor::execute` now checks the canonical `decision_hash` key first: if it already holds a **terminal** record, the canonical record wins and the legacy receipt key is not adopted — so a stale non-terminal legacy row no longer causes a re-dispatch when the canonical decision is already done. The legacy key is still adopted for the same decision when the canonical key is absent or non-terminal (terminal-legacy dedupe and in-flight-legacy continuity unchanged).

Added a replay test seeding both a terminal canonical record and a stale non-terminal legacy record; the existing legacy-replay, in-flight, distinct-decision, and forward tests are unchanged and still pass.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: icn/crates/icn-core/src/supervisor/decision_executor.rs [hunk 6]
```diff
@@ -391,6 +391,49 @@ impl DecisionExecutor {
...
+        // When BOTH rows exist for the same decision (reachable only if a node
+        // ran an intermediate hash-keying build before this compat fix landed),
+        // the canonical hash-keyed record wins if it is already terminal: never
+        // re-dispatch under a stale legacy row when the canonical decision is done.
+        let canonical_is_terminal = self
+            .store
+            .get(decision_hash)?
+            .map(|r| r.is_terminal())
+            .unwrap_or(false);
+        let decision_hash: &str = if decision_hash != decision_receipt_id && !canonical_is_terminal
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs:423]: Addressed in c0aa92cf. `execute` now also guards the recovery case: when `decision_hash == decision_receipt_id` (a legacy receipt-keyed row re-executed by `recover_in_flight` under its stored key, which bypasses the callback-path shadow), it derives the federation `decision_hash` from the stored effects and dedupes if that canonical record is already terminal. The canonical-terminal-wins invariant now holds at **both** `execute` entry points (the executor callback and `recover_in_flight`).

Added a recovery-path test seeding a terminal canonical record plus a stale `Executing` legacy row; `recover_in_flight` must not re-dispatch the legacy row.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: icn/crates/icn-core/src/supervisor/decision_executor.rs [hunk 7]
```diff
@@ -391,6 +391,75 @@ impl DecisionExecutor {
...
+        {
+            match self.store.get(decision_receipt_id)? {
+                Some(legacy)
+                    if extract_decision_hash(&legacy.effects).as_deref() == Some(decision_hash) =>
+                {
+                    decision_receipt_id
```
[the remark this answers is not in the document]
Comment (fahertym) [reply icn/crates/icn-core/src/supervisor/decision_executor.rs:429]: Tracked as #2102 rather than fixed in this PR. The "both rows exist" state (a canonical hash-keyed row alongside a legacy receipt-keyed row for the same decision) is unreachable on any merged/released build — it requires a node to have run an un-merged intermediate #2096 commit (`cbc48fc2`/`83acaeef`/`b2b53446`) then upgraded. The compat shim here is deliberately **read-only** key selection; the reconciliation you describe (mark/delete/alias the stale canonical row) is record mutation that belongs in its own scoped PR. The reachable `#2094 → #2096` boundary — terminal/in-flight/distinct-decision compat plus canonical-terminal-wins at both `execute` entry points — is handled in this PR. Defense-in-depth follow-up: #2102.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
